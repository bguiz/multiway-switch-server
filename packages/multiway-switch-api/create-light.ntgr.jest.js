'use strict';

const serverTest = require('../apis/server-test.js');

// LET THERE BE LIGHT!!
describe('[create light]', () => {

	it('should create a new light with switches', (done) => {
		serverTest.getRequest()
			.get('/api/multiway-switch/v1/create-light')
			.query({
				lightId: 1,
				numSwitches: 2,
			})
			.send()
			.end((err, response) => {
				expect(err).not.toEqual(expect.anything());
				expect({
					status: response.status,
					body: response.body,
				}).toMatchSnapshot();
				done();
			});
	});

	//FIXME add this - persistence test

	it('should have persisted newly created light', (done) => {
		const persist = require('./persist.js');
		const existingLight = persist.get('light', 1);
		expect(existingLight).toEqual(expect.anything());
		expect(existingLight).toMatchSnapshot();
		done();
	});

	// Cover the error cases

	//FIXME add these - error code coverage
	it('should not create a new light when number of switches is unsupported', (done) => {
		serverTest.getRequest()
			.get('/api/multiway-switch/v1/create-light')
			.query({
				lightId: 2,
				numSwitches: 3,
			})
			.send()
			.end((err, response) => {
				expect(err).not.toEqual(expect.anything());
				expect({
					status: response.status,
					body: response.body,
				}).toMatchSnapshot();
				done();
			});
	});

	it('should not create a new light ID has conflict', (done) => {
		serverTest.getRequest()
			.get('/api/multiway-switch/v1/create-light')
			.query({
				lightId: 1,
				numSwitches: 2,
			})
			.send()
			.end((err, response) => {
				expect(err).not.toEqual(expect.anything());
				expect({
					status: response.status,
					body: response.body,
				}).toMatchSnapshot();
				done();
			});
	});

	// parametric should not create new lights based on numSwitches

	//FIXME add these - parametric tests
	[-2,0,1,3,10, 2].forEach((numSwitches, idx) => {

		it(`should not create a new light with ${numSwitches} switches`, (done) => {
			const lightId = 100 + idx;
			serverTest.getRequest()
				.get('/api/multiway-switch/v1/create-light')
				.query({
					lightId,
					numSwitches,
				})
				.send()
				.end((err, response) => {
					expect(err).not.toEqual(expect.anything());
					expect({
						status: response.status,
						body: response.body,
					}).toMatchSnapshot();
					done();
				});
		});

	});

});
