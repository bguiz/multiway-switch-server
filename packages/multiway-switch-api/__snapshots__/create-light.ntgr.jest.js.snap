// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[create light] should create a new light with switches 1`] = `
{
  "body": {
    "lightId": 1,
    "lightState": "OFF",
    "switchStates": [
      "UP",
      "UP"
    ]
  },
  "status": 200
}
`;

exports[`[create light] should have persisted newly created light 1`] = `
{
  "switchStates": [
    false,
    false
  ]
}
`;

exports[`[create light] should not create a new light ID has conflict 1`] = `
{
  "body": {
    "code": 400002,
    "message": "Light with this ID already exists."
  },
  "status": 400
}
`;

exports[`[create light] should not create a new light when number of switches is unsupported 1`] = `
{
  "body": {
    "code": 400001,
    "message": "Only supports 2 switches."
  },
  "status": 400
}
`;

exports[`[create light] should not create a new light with -2 switches 1`] = `
{
  "body": {
    "code": 400001,
    "message": "Only supports 2 switches."
  },
  "status": 400
}
`;

exports[`[create light] should not create a new light with 0 switches 1`] = `
{
  "body": {
    "code": 400001,
    "message": "Only supports 2 switches."
  },
  "status": 400
}
`;

exports[`[create light] should not create a new light with 1 switches 1`] = `
{
  "body": {
    "code": 400001,
    "message": "Only supports 2 switches."
  },
  "status": 400
}
`;

exports[`[create light] should not create a new light with 2 switches 1`] = `
{
  "body": {
    "lightId": 105,
    "lightState": "OFF",
    "switchStates": [
      "UP",
      "UP"
    ]
  },
  "status": 200
}
`;

exports[`[create light] should not create a new light with 3 switches 1`] = `
{
  "body": {
    "code": 400001,
    "message": "Only supports 2 switches."
  },
  "status": 400
}
`;

exports[`[create light] should not create a new light with 10 switches 1`] = `
{
  "body": {
    "code": 400001,
    "message": "Only supports 2 switches."
  },
  "status": 400
}
`;
